dotnet/src/Connectors/Connectors.Memory.AzureAISearch/AzureAISearchGenericDataModelMapper.cs

```diff
@@ -0,0 +1,177 @@
+
+                    var serializedJsonNode = JsonSerializer.SerializeToNode(dataModel.Vectors[vectorProperty.DataModelPropertyName]);
+                    storageJsonObject.Add(storagePropertyName, serializedJsonNode);
+                }
+            }
+        }
+
+        return storageJsonObject;
+    }
+
+    /// <inheritdoc />
+    public VectorStoreGenericDataModel<string> MapFromStorageToDataModel(JsonObject storageModel, StorageToDataModelMapperOptions options)
+    {
+        Verify.NotNull(storageModel);
+
+        // Create variables to store the response properties.
+        var dataProperties = new Dictionary<string, object?>();
+        var vectorProperties = new Dictionary<string, object?>();
+        string key = string.Empty;
+
+        // Loop through all known properties and map each from json to the data type.
+        foreach (var property in this._vectorStoreRecordDefinition.Properties)
+        {
+            if (property is VectorStoreRecordKeyProperty keyProperty)
+            {
+                var storagePropertyName = keyProperty.StoragePropertyName ?? keyProperty.DataModelPropertyName;
+                var value = storageModel[storagePropertyName];
+                if (value == null)
+                {
+                    throw new InvalidOperationException($"The key property '{storagePropertyName}' is missing from the record retrieved from storage.");
+                }
+
+                key = (string)value!;
+            }
+            else if (property is VectorStoreRecordDataProperty dataProperty)
+            {
+                var storagePropertyName = dataProperty.StoragePropertyName ?? dataProperty.DataModelPropertyName;
+                var value = storageModel[storagePropertyName];
+                if (value != null)
+                {
+                    dataProperties.Add(dataProperty.DataModelPropertyName, GetDataPropertyValue(property.PropertyType, value));
+                }
+                else
+                {
+                    dataProperties.Add(dataProperty.DataModelPropertyName, null);
+                }
+            }
+            else if (property is VectorStoreRecordVectorProperty vectorProperty && options.IncludeVectors)
+            {
+                var storagePropertyName = vectorProperty.StoragePropertyName ?? vectorProperty.DataModelPropertyName;
+                var value = storageModel[storagePropertyName];
+                if (value != null)
+                {
+                    ReadOnlyMemory<float> vector = value.AsArray().Select(x => (float)x!).ToArray();
+                    vectorProperties.Add(vectorProperty.DataModelPropertyName, vector);
+                }
+                else
+                {
+                    vectorProperties.Add(vectorProperty.DataModelPropertyName, null);
+                }
+            }
+        }
+
+        return new VectorStoreGenericDataModel<string>(key) { Data = dataProperties, Vectors = vectorProperties };
```

Comment:
Do we need to test for a to make sure at a minimum a key value was set?

Reply:
Updated key to nullable and added a check to make sure it's provided at the end.
That said, there is a lot of validation to make sure that a key is specified.  There may be cases though where there a key defined but it didn't exist in the retrieved record, so adding a check makes sense.